File: src/main/java/com/ggardet/patient/internal/domain/PatientServiceImpl.java
package com.ggardet.patient.internal.domain;

import com.ggardet.patient.internal.entity.Patient;
import com.ggardet.patient.internal.repository.PatientRepository;
import com.ggardet.patient.spi.PatientService;
import com.ggardet.patient.spi.dto.PatientDTO;
import com.ggardet.patient.spi.exception.PatientNotFoundException;
import com.ggardet.patient.spi.mapper.PatientMapper;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;
import java.util.Objects;
import java.util.UUID;

@Service
@RequiredArgsConstructor
@Transactional
public class PatientServiceImpl implements PatientService {
    private final PatientRepository patientRepository;
    private final PatientMapper patientMapper;

    @Override
    public PatientDTO create(final PatientDTO patientDTO) {
        final var patient = patientMapper.toEntity(patientDTO);
        final var savedPatient = patientRepository.save(patient);
        return patientMapper.toDto(savedPatient);
    }

    @Override
    @Transactional(readOnly = true)
    public PatientDTO findById(final UUID id) {
        final var patient = patientRepository.findById(id)
                .orElseThrow(() -> new PatientNotFoundException(id));
        return patientMapper.toDto(patient);
    }

    @Override
    @Transactional(readOnly = true)
    public List<PatientDTO> findAll() {
        final var patients = patientRepository.findAll();
        return patientMapper.toDtos(patients);
    }

    @Override
    @Transactional(readOnly = true)
    public List<PatientDTO> findByHospitalId(final UUID hospitalId) {
        final var patients = patientRepository.findByHospitalId(hospitalId);
        return patientMapper.toDtos(patients);
    }

    @Override
    @Transactional(readOnly = true)
    public List<PatientDTO> findByServiceId(final UUID serviceId) {
        final var patients = patientRepository.findByServiceId(serviceId);
        return patientMapper.toDtos(patients);
    }

    @Override
    @Transactional(readOnly = true)
    public PatientDTO findBySocialSecurityNumber(final String socialSecurityNumber) {
        final var patient = patientRepository.findBySocialSecurityNumber(socialSecurityNumber)
                .orElseThrow(() -> new PatientNotFoundException(null));
        return patientMapper.toDto(patient);
    }

    @Override
    @Transactional(readOnly = true)
    public List<PatientDTO> searchByName(final String name) {
        final var patients = patientRepository.findByFirstnameContainingIgnoreCaseOrLastnameContainingIgnoreCase(name, name);
        return patientMapper.toDtos(patients);
    }

    @Override
    public PatientDTO update(final UUID id, final PatientDTO patientDTO) {
        final var existingPatient = patientRepository.findById(id)
                .orElseThrow(() -> new PatientNotFoundException(id));
        updatePatientFields(existingPatient, patientDTO);
        final var updatedPatient = patientRepository.save(existingPatient);
        return patientMapper.toDto(updatedPatient);
    }

    @Override
    public void deleteById(final UUID id) {
        if (!patientRepository.existsById(id)) {
            throw new PatientNotFoundException(id);
        }
        patientRepository.deleteById(id);
    }

    private void updatePatientFields(final Patient existingPatient, final PatientDTO dto) {
        if (Objects.nonNull(dto.firstname())) {
            existingPatient.setFirstname(dto.firstname());
        }
        if (Objects.nonNull(dto.lastname())) {
            existingPatient.setLastname(dto.lastname());
        }
        if (Objects.nonNull(dto.birthDate())) {
            existingPatient.setBirthDate(dto.birthDate());
        }
        if (Objects.nonNull(dto.socialSecurityNumber())) {
            existingPatient.setSocialSecurityNumber(dto.socialSecurityNumber());
        }
        if (Objects.nonNull(dto.phoneNumber())) {
            existingPatient.setPhoneNumber(dto.phoneNumber());
        }
        if (Objects.nonNull(dto.email())) {
            existingPatient.setEmail(dto.email());
        }
        if (Objects.nonNull(dto.hospitalId())) {
            existingPatient.setHospitalId(dto.hospitalId());
        }
        if (Objects.nonNull(dto.serviceId())) {
            existingPatient.setServiceId(dto.serviceId());
        }
    }
}

File: src/main/kotlin/com/ggardet/patient/internal/domain/PatientServiceImpl.kt
package com.ggardet.patient.internal.domain

import com.ggardet.patient.internal.entity.Patient
import com.ggardet.patient.internal.repository.PatientRepository
import com.ggardet.patient.spi.PatientService
import com.ggardet.patient.spi.dto.PatientDTO
import com.ggardet.patient.spi.exception.PatientNotFoundException
import com.ggardet.patient.spi.mapper.PatientMapper
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
@Transactional
class PatientServiceImpl(
    private val patientRepository: PatientRepository,
    private val patientMapper: PatientMapper
) : PatientService {

    override fun create(patientDTO: PatientDTO): PatientDTO {
        val patient = patientMapper.toEntity(patientDTO)
        val savedPatient = patientRepository.save(patient)
        return patientMapper.toDto(savedPatient)
    }

    @Transactional(readOnly = true)
    override fun findById(id: UUID): PatientDTO {
        val patient = patientRepository.findById(id)
            .orElseThrow { PatientNotFoundException(id) }
        return patientMapper.toDto(patient)
    }

    @Transactional(readOnly = true)
    override fun findAll(): List<PatientDTO> {
        val patients = patientRepository.findAll()
        return patientMapper.toDtos(patients)
    }

    @Transactional(readOnly = true)
    override fun findByHospitalId(hospitalId: UUID): List<PatientDTO> {
        val patients = patientRepository.findByHospitalId(hospitalId)
        return patientMapper.toDtos(patients)
    }

    @Transactional(readOnly = true)
    override fun findByServiceId(serviceId: UUID): List<PatientDTO> {
        val patients = patientRepository.findByServiceId(serviceId)
        return patientMapper.toDtos(patients)
    }

    @Transactional(readOnly = true)
    override fun findBySocialSecurityNumber(socialSecurityNumber: String): PatientDTO {
        val patient = patientRepository.findBySocialSecurityNumber(socialSecurityNumber)
            .orElseThrow { PatientNotFoundException(null) }
        return patientMapper.toDto(patient)
    }

    @Transactional(readOnly = true)
    override fun searchByName(name: String): List<PatientDTO> {
        val patients = patientRepository.findByFirstnameContainingIgnoreCaseOrLastnameContainingIgnoreCase(name, name)
        return patientMapper.toDtos(patients)
    }

    override fun update(id: UUID, patientDTO: PatientDTO): PatientDTO {
        val existingPatient = patientRepository.findById(id)
            .orElseThrow { PatientNotFoundException(id) }
        updatePatientFields(existingPatient, patientDTO)
        val updatedPatient = patientRepository.save(existingPatient)
        return patientMapper.toDto(updatedPatient)
    }

    override fun deleteById(id: UUID) {
        if (!patientRepository.existsById(id)) {
            throw PatientNotFoundException(id)
        }
        patientRepository.deleteById(id)
    }

    private fun updatePatientFields(existingPatient: Patient, dto: PatientDTO) {
        dto.firstname.let { existingPatient.firstname = it }
        dto.lastname.let { existingPatient.lastname = it }
        existingPatient.birthDate = dto.birthDate
        dto.socialSecurityNumber?.let { existingPatient.socialSecurityNumber = it }
        dto.phoneNumber?.let { existingPatient.phoneNumber = it }
        dto.email?.let { existingPatient.email = it }
        dto.hospitalId?.let { existingPatient.hospitalId = it }
        dto.serviceId?.let { existingPatient.serviceId = it }
    }
}